plugins/gssapi.c

```diff
@@ -877,15 +1118,48 @@ gssapi_server_mech_authneg(context_t *text,
 
 	/* If caller didn't provide creds already */
 	if ( server_creds == GSS_C_NO_CREDENTIAL) {
+#ifdef GSS_USE_LOAD_FROM
+            const sasl_utils_t *utils = text->utils;
+            unsigned int rl; /* Required storage by getopt() */
+            char* keytab = NULL; /* The configured credentials cache file path */
+            utils->getopt(utils->getopt_context, "GSSAPI", "keytab", (const char**)&keytab, &rl);
+#endif
 	    GSS_LOCK_MUTEX_CTX(params->utils, text);
-	    maj_stat = gss_acquire_cred(&min_stat, 
-					text->server_name,
-					GSS_C_INDEFINITE, 
-					GSS_C_NO_OID_SET,
-					GSS_C_ACCEPT,
-					&text->server_creds, 
-					NULL, 
-					NULL);
+#ifdef GSS_USE_LOAD_FROM
+            if (keytab != NULL) { /* Force cached credentials to use GSS Proxy compliant API of GSS */
+                gss_key_value_element_desc ccache_element = {.key = "ccache", .value = NULL};
```

Comment:
Why are you setting ccache to NULL here?
I know NULL values currently work ( I wrote this API) but they are an undefined behavior and should not be used.
It may crash the program in the future.

Reply:
Right... Hm. I have mostly (only) seen this kind of approach in other project just about everywhere. I just suspected it would do some kind of fallback. Any good ideas on how to make it correct? I don't think it should be used at all this early (since it's basically overridden once a user has authenticated by the other code).

Also, I know it's a bit awekward to load the service creds like this (when a user connects), i know. But for now it will do. - Could be up for a move to program initialization at some point in the future.

Reply:
I guess the "ccache" key/value (or rather KRB5_CS_CCACHE_URN) could be omitted altogether. *thinking*

Reply:
There was also a weird assignment that got immediately overridden there... Cleaning this up a bit. Good eye!

Reply:
Just drop setting the ccache value if you do not want to set it to a specific value, the NULL value, while currently accepted by the MIT implementation is an invalid input.

Reply:
Yeah, i think the copy'n'paste monster took a dump in there :) Making it so.

Just need to run a regression run before pushing.

Reply:
Sorry for the delay... Got stuck with it for a while. Something seems borked on the client side from either 7a95382c68e7187fe7407b2a94036d9ca3246e34 or 238380260fe623212c0f21d63e763b7a849540d1 - At least I cannot test like I did before (I initially based my work on the latest release, now it's branched from master), by just initializing a default principal / ticket cache with kinit and then start the sample-client. Have to use the last released one, client side, where it works, for testing. Took me a while to figure out what bailed on me tho. @simo5 , see any reason why? It does not feel like an odd use-case. Just using the default available token to authenticate using GSSAPI :) It now exits since gss_acquire_cred() fails, or maybe as a consequence for gsskrb5_register_acceptor_identity() failure. (This is however unrelated to this pull-request besides the fact that it seem to complicate use of GSSAPI for next release - or more likely; I'm misunderstanding something).

Reply:
In general gsskrb5_register_acceptor_identity() should be avoided as it is basically an irreversible operation, and it is not needed with modern APIs like load_from/store_into you are trying to use here.

Also why those patches limit the path to 1024 chars in 2019? We have PATH_MAX available if really needed (which is defined as 4096 on linux for example), and in general allocating on the heap would be preferred, no point in wasting 1k of stack like that ... but I digress.

Please do make gss_acquire_cred_from mutually exclusive with gssapi_register_acceptor_identity() with the former being the preferred feature to use and register identity only as an optional fallback.

Reply:
Right... Just to be clear.... I did add gss_acquire_cred_from server-side. The problem I described above with the client side is a contribution by someone else, back in July, but it makes it (for me) "un-possible" to test my server-side changes. With what's on master/my fork right now :)

Reply:
@simo5 yay! #591 will fix the issue with not reading the default creds. However just getting _this_ pull request into master will make the delegation feature-incomplete. Both are needed.